Test open data row deletion fix

wrridgeway opened the pull request:
High-level summary: This PR grabs all rows that have been deactivated or re-classed to 999 (parcel status view allows 999s since they are explicitly indicated in the view) in pardat (and occasionally other tables that define a view's universe) and outer joins them to `open_data` views so that uploads to socrata will include all rows that should not be present on the open data portal.

Note: this update will break our current open data testing syntax. we will have to make sure to count only rows from open_data athena views `where ":deleted" is null`

Pay special attention to the sf/mf view. sf/mf now removes 999s and deactivated `pardat` parcels at the `default` view level which were sneaking in previously, so there is a slight disagreement with the old open data view. For this view I've compared the new open data view to the new `default.vw_card_res_char`, instead of the old open data view. [Otherwise, the open data test workflow is returning what we'd hope for it to return.](https://github.com/ccao-data/data-architecture/actions/runs/15000276948/job/42147705471)

<details>

<summary>check row_id uniqueness</summary>

```
select 'vw_assessed_value' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_assessed_value
group by row_id
having count(*) > 1

union all

select 'vw_parcel_address' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_address
group by row_id
having count(*) > 1

union all

select 'w_parcel_status' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_status
group by row_id
having count(*) > 1

union all

select 'vw_parcel_universe_current' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_current
group by row_id
having count(*) > 1

union all

select 'vw_parcel_universe_historical' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_historical
group by row_id
having count(*) > 1

union all

select 'vw_property_tax_exempt_parcel' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_property_tax_exempt_parcel
group by row_id
having count(*) > 1

union all

select 'vw_res_condo_unit_char' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_res_condo_unit_char
group by row_id
having count(*) > 1

union all

select 'vw_sf_mf_improvement_char' as view_name, row_id, count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_sf_mf_improvement_char
group by row_id
having count(*) > 1
```

returns 0 rows

</details>

These test queries are more complex than you'd hope since you can't join on NULL values of `:deleted`.

<details>

<summary>vw_assessed_value</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_assessed_value
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_assessed_value
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 4  | true     | 2025 | 2162    |         |       |
| 5  | true     | 2021 | 3229    |         |       |
| 10 | true     | 2024 | 4130    |         |       |
| 25 | true     | 2022 | 4782    |         |       |
| 31 | true     | 2023 | 5896    |         |       |
| 1  |          | 2022 | 1865013 | 1865013 | true  |
| 2  |          | 2021 | 1864589 | 1864589 | true  |
| 3  |          | 2001 | 1655168 | 1655168 | true  |
| 6  |          | 2020 | 1864035 | 1864035 | true  |
| 7  |          | 2024 | 1864107 | 1864107 | true  |
| 8  |          | 2002 | 1673530 | 1673530 | true  |
| 9  |          | 2012 | 1862181 | 1862181 | true  |
| 11 |          | 2009 | 1849650 | 1849650 | true  |
| 12 |          | 2016 | 1863996 | 1863996 | true  |
| 13 |          | 2007 | 1804217 | 1804217 | true  |
| 14 |          | 2011 | 1861008 | 1861008 | true  |
| 15 |          | 2018 | 1864588 | 1864588 | true  |
| 16 |          | 2003 | 1693139 | 1693139 | true  |
| 17 |          | 2006 | 1771313 | 1771313 | true  |
| 18 |          | 2023 | 1864161 | 1864161 | true  |
| 19 |          | 2019 | 1864102 | 1864102 | true  |
| 20 |          | 2025 | 1863564 | 1863564 | true  |
| 21 |          | 2017 | 1864621 | 1864621 | true  |
| 22 |          | 2013 | 1861929 | 1861929 | true  |
| 23 |          | 1999 | 1620798 | 1620798 | true  |
| 24 |          | 2010 | 1859244 | 1859244 | true  |
| 26 |          | 2015 | 1862756 | 1862756 | true  |
| 27 |          | 2000 | 1637079 | 1637079 | true  |
| 28 |          | 2004 | 1716214 | 1716214 | true  |
| 29 |          | 2014 | 1862170 | 1862170 | true  |
| 30 |          | 2008 | 1832617 | 1832617 | true  |
| 32 |          | 2005 | 1740963 | 1740963 | true  |

</details>

<details>

<summary>vw_parcel_address</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_address
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_parcel_address
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 1  | true     | 2023 | 5896    |         |       |
| 20 | true     | 2022 | 4782    |         |       |
| 28 | true     | 2025 | 2162    |         |       |
| 29 | true     | 2021 | 3229    |         |       |
| 31 | true     | 2024 | 4130    |         |       |
| 2  |          | 2021 | 1864589 | 1864589 | true  |
| 3  |          | 2015 | 1862756 | 1862756 | true  |
| 4  |          | 2001 | 1655398 | 1655398 | true  |
| 5  |          | 2017 | 1864621 | 1864621 | true  |
| 6  |          | 2003 | 1693153 | 1693153 | true  |
| 7  |          | 2020 | 1864035 | 1864035 | true  |
| 8  |          | 2025 | 1863564 | 1863564 | true  |
| 9  |          | 2002 | 1673735 | 1673735 | true  |
| 10 |          | 2009 | 1849650 | 1849650 | true  |
| 11 |          | 2011 | 1861008 | 1861008 | true  |
| 12 |          | 2007 | 1804219 | 1804219 | true  |
| 13 |          | 2004 | 1716218 | 1716218 | true  |
| 14 |          | 1999 | 1620990 | 1620990 | true  |
| 15 |          | 2018 | 1864590 | 1864590 | true  |
| 16 |          | 2005 | 1740965 | 1740965 | true  |
| 17 |          | 2022 | 1865016 | 1865016 | true  |
| 18 |          | 2019 | 1864102 | 1864102 | true  |
| 19 |          | 2000 | 1637269 | 1637269 | true  |
| 21 |          | 2013 | 1861935 | 1861935 | true  |
| 22 |          | 2016 | 1863996 | 1863996 | true  |
| 23 |          | 2006 | 1771314 | 1771314 | true  |
| 24 |          | 2012 | 1862181 | 1862181 | true  |
| 25 |          | 2023 | 1864161 | 1864161 | true  |
| 26 |          | 2014 | 1862170 | 1862170 | true  |
| 27 |          | 2008 | 1832618 | 1832618 | true  |
| 30 |          | 2024 | 1864107 | 1864107 | true  |
| 32 |          | 2010 | 1859244 | 1859244 | true  |

</details>

<details>

<summary>vw_parcel_status</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_status
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_parcel_status
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 10 | true     | 2021 | 3229    |         |       |
| 11 | true     | 2025 | 1557    |         |       |
| 19 | true     | 2023 | 5895    |         |       |
| 23 | true     | 2024 | 4130    |         |       |
| 28 | true     | 2022 | 4780    |         |       |
| 1  |          | 2005 | 1740965 | 1740965 | true  |
| 2  |          | 2022 | 1865018 | 1865018 | true  |
| 3  |          | 2001 | 1655398 | 1655398 | true  |
| 4  |          | 2016 | 1863996 | 1863996 | true  |
| 5  |          | 2007 | 1804219 | 1804219 | true  |
| 6  |          | 1999 | 1620990 | 1620990 | true  |
| 7  |          | 2017 | 1864621 | 1864621 | true  |
| 8  |          | 2003 | 1693153 | 1693153 | true  |
| 9  |          | 2020 | 1864035 | 1864035 | true  |
| 12 |          | 2012 | 1862181 | 1862181 | true  |
| 13 |          | 2025 | 1864169 | 1864169 | true  |
| 14 |          | 2023 | 1864162 | 1864162 | true  |
| 15 |          | 2004 | 1716218 | 1716218 | true  |
| 16 |          | 2018 | 1864590 | 1864590 | true  |
| 17 |          | 2006 | 1771314 | 1771314 | true  |
| 18 |          | 2015 | 1862756 | 1862756 | true  |
| 20 |          | 2021 | 1864589 | 1864589 | true  |
| 21 |          | 2002 | 1673735 | 1673735 | true  |
| 22 |          | 2009 | 1849650 | 1849650 | true  |
| 24 |          | 2024 | 1864107 | 1864107 | true  |
| 25 |          | 2019 | 1864102 | 1864102 | true  |
| 26 |          | 2000 | 1637269 | 1637269 | true  |
| 27 |          | 2010 | 1859244 | 1859244 | true  |
| 29 |          | 2014 | 1862170 | 1862170 | true  |
| 30 |          | 2008 | 1832618 | 1832618 | true  |
| 31 |          | 2013 | 1861935 | 1861935 | true  |
| 32 |          | 2011 | 1861008 | 1861008 | true  |

</details>

<details>

<summary>vw_parcel_universe_current</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_current
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_parcel_universe_current
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| # | :deleted | year | new     | old     | match |
|---|----------|------|---------|---------|-------|
| 1 | true     | 2025 | 2162    |         |       |
| 2 |          | 2025 | 1863564 | 1863564 | true  |

</details>

<details>

<summary>vw_parcel_universe_historical</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_historical
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_parcel_universe_historical
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 19 | true     | 2025 | 2162    |         |       |
| 20 | true     | 2021 | 3229    |         |       |
| 24 | true     | 2023 | 5896    |         |       |
| 26 | true     | 2024 | 4130    |         |       |
| 27 | true     | 2022 | 4782    |         |       |
| 1  |          | 2020 | 1864035 | 1864035 | true  |
| 2  |          | 2003 | 1693153 | 1693153 | true  |
| 3  |          | 2009 | 1849650 | 1849650 | true  |
| 4  |          | 2001 | 1655398 | 1655398 | true  |
| 5  |          | 2022 | 1865016 | 1865016 | true  |
| 6  |          | 2018 | 1864590 | 1864590 | true  |
| 7  |          | 2021 | 1864589 | 1864589 | true  |
| 8  |          | 2017 | 1864621 | 1864621 | true  |
| 9  |          | 2006 | 1771314 | 1771314 | true  |
| 10 |          | 2002 | 1673735 | 1673735 | true  |
| 11 |          | 2013 | 1861935 | 1861935 | true  |
| 12 |          | 2012 | 1862181 | 1862181 | true  |
| 13 |          | 2005 | 1740965 | 1740965 | true  |
| 14 |          | 2016 | 1863996 | 1863996 | true  |
| 15 |          | 2004 | 1716218 | 1716218 | true  |
| 16 |          | 2010 | 1859244 | 1859244 | true  |
| 17 |          | 1999 | 1620990 | 1620990 | true  |
| 18 |          | 2025 | 1863564 | 1863564 | true  |
| 21 |          | 2007 | 1804219 | 1804219 | true  |
| 22 |          | 2011 | 1861008 | 1861008 | true  |
| 23 |          | 2015 | 1862756 | 1862756 | true  |
| 25 |          | 2024 | 1864107 | 1864107 | true  |
| 28 |          | 2014 | 1862170 | 1862170 | true  |
| 29 |          | 2008 | 1832618 | 1832618 | true  |
| 30 |          | 2023 | 1864161 | 1864161 | true  |
| 31 |          | 2000 | 1637269 | 1637269 | true  |
| 32 |          | 2019 | 1864102 | 1864102 | true  |

</details>

<details>

<summary>vw_property_tax_exempt_parcel</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_property_tax_exempt_parcel
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_property_tax_exempt_parcel
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| # | :deleted | year | new   | old   | match |
|---|----------|------|-------|-------|-------|
| 2 | true     | 2023 | 5896  |       |       |
| 3 | true     | 2025 | 2162  |       |       |
| 4 | true     | 2021 | 3229  |       |       |
| 7 | true     | 2022 | 4782  |       |       |
| 9 | true     | 2024 | 4130  |       |       |
| 1 |          | 2024 | 93509 | 93509 | true  |
| 5 |          | 2023 | 92481 | 92481 | true  |
| 6 |          | 2025 | 93409 | 93409 | true  |
| 8 |          | 2022 | 92416 | 92416 | true  |

</details>

<details>

<summary>vw_res_condo_unit_char</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_res_condo_unit_char
group by year, ":deleted"
),

old AS (
select false as ":deleted", year, count(*) as old
from open_data.vw_res_condo_unit_char
group by year, false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new    | old    | match |
|----|----------|------|--------|--------|-------|
| 1  | true     | 2023 | 5896   |        |       |
| 7  | true     | 2025 | 2162   |        |       |
| 8  | true     | 2021 | 3229   |        |       |
| 10 | true     | 2024 | 4130   |        |       |
| 21 | true     | 2022 | 4782   |        |       |
| 2  |          | 1999 | 235795 | 235795 | true  |
| 3  |          | 2020 | 458449 | 458449 | true  |
| 4  |          | 2001 | 266830 | 266830 | true  |
| 5  |          | 2006 | 370734 | 370734 | true  |
| 6  |          | 2016 | 458847 | 458847 | true  |
| 9  |          | 2012 | 458194 | 458194 | true  |
| 11 |          | 2004 | 321082 | 321082 | true  |
| 12 |          | 2024 | 457315 | 457315 | true  |
| 13 |          | 2002 | 283450 | 283450 | true  |
| 14 |          | 2022 | 459126 | 459126 | true  |
| 15 |          | 2013 | 457812 | 457812 | true  |
| 16 |          | 2003 | 300884 | 300884 | true  |
| 17 |          | 2018 | 459473 | 459473 | true  |
| 18 |          | 2011 | 457166 | 457166 | true  |
| 19 |          | 2009 | 445790 | 445790 | true  |
| 20 |          | 2014 | 458049 | 458049 | true  |
| 22 |          | 2010 | 454939 | 454939 | true  |
| 23 |          | 2005 | 342996 | 342996 | true  |
| 24 |          | 2023 | 457640 | 457640 | true  |
| 25 |          | 2017 | 459275 | 459275 | true  |
| 26 |          | 2021 | 458980 | 458980 | true  |
| 27 |          | 2007 | 402508 | 402508 | true  |
| 28 |          | 2025 | 457216 | 457216 | true  |
| 29 |          | 2008 | 429185 | 429185 | true  |
| 30 |          | 2015 | 458355 | 458355 | true  |
| 31 |          | 2000 | 250302 | 250302 | true  |
| 32 |          | 2019 | 458673 | 458673 | true  |

</details>

<details>

<summary>vw_sf_mf_improvement_char</summary>

```
with new as (
select case when ":deleted" then true else false end as ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_sf_mf_improvement_char
group by year, ":deleted"
),

old AS (
select false as ":deleted", cast(year as int) as year, count(*) as old
from z_ci_fix_open_data_row_deletion_default.vw_card_res_char
group by cast(year as int), false
)

select
    case when new.":deleted" then true else null end as ":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old
on new.year = old.year
and new.":deleted"= old.":deleted"
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 11 | true     | 2021 | 1410    |         |       |
| 12 | true     | 2025 | 798     |         |       |
| 23 | true     | 2024 | 1803    |         |       |
| 25 | true     | 2020 | 6       |         |       |
| 27 | true     | 2023 | 1667    |         |       |
| 29 | true     | 2022 | 1351    |         |       |
| 1  |          | 2023 | 1101126 | 1101126 | true  |
| 2  |          | 2020 | 1099085 | 1099085 | true  |
| 3  |          | 1999 | 1059702 | 1059702 | true  |
| 4  |          | 2007 | 1098544 | 1098544 | true  |
| 5  |          | 2019 | 1098966 | 1098966 | true  |
| 6  |          | 2017 | 1098637 | 1098637 | true  |
| 7  |          | 2022 | 1100239 | 1100239 | true  |
| 8  |          | 2002 | 1081904 | 1081904 | true  |
| 9  |          | 2014 | 1097705 | 1097705 | true  |
| 10 |          | 2016 | 1098108 | 1098108 | true  |
| 13 |          | 2009 | 1099466 | 1099466 | true  |
| 14 |          | 2006 | 1096341 | 1096341 | true  |
| 15 |          | 2013 | 1098025 | 1098025 | true  |
| 16 |          | 2021 | 1099631 | 1099631 | true  |
| 17 |          | 2011 | 1099010 | 1099010 | true  |
| 18 |          | 2000 | 1077974 | 1077974 | true  |
| 19 |          | 2015 | 1098115 | 1098115 | true  |
| 20 |          | 2004 | 1089762 | 1089762 | true  |
| 21 |          | 2012 | 1098752 | 1098752 | true  |
| 22 |          | 2010 | 1099597 | 1099597 | true  |
| 24 |          | 2005 | 1093404 | 1093404 | true  |
| 26 |          | 2003 | 1085834 | 1085834 | true  |
| 28 |          | 2018 | 1098786 | 1098786 | true  |
| 30 |          | 2001 | 1078951 | 1078951 | true  |
| 31 |          | 2024 | 1102434 | 1102434 | true  |
| 32 |          | 2008 | 1099960 | 1099960 | true  |
| 33 |          | 2025 | 1102311 | 1102311 | true  |

</details>

wrridgeway commented on socrata/socrata_upload.py:
```diff
@@ -236,7 +239,7 @@ def build_query_dict(athena_asset, asset_id, years=None):
     print(columns)
 
-    query = f"SELECT {', '.join(columns['column'])} FROM {athena_asset}"
+    query = f"""SELECT {", ".join(columns["column"])}, ":deleted" FROM {athena_asset}"""
```
This ensures the `:deleted` field is included in athena query and is uploaded. I think this should work across all views.

wrridgeway commented on socrata/socrata_upload.py:
```diff
@@ -98,6 +98,9 @@ def parse_assets(assets=None):
         assets = [asset.strip() for asset in str(assets).split(",")]
 
         all_assets = all_assets[all_assets["label"].isin(assets)]
+        all_assets["athena_asset"] = all_assets["athena_asset"].str.replace(
+            "open_data", "z_ci_fix_open_data_row_deletion_open_data"
+        )
```
Temporary, won't be merged.

jeancochrane replied:
**[Nitpick, required]** Reminder to switch this back before we merge.

wrridgeway commented on dbt/models/open_data/exposures.yml:
```diff
@@ -137,7 +137,7 @@ exposures:
     meta:
       test_row_count: true
-      asset_id: pabr-t5kh
+      asset_id: y7q2-6pvn
```
Temporary, won't be merged.

wrridgeway commented on dbt/models/open_data/open_data.vw_parcel_universe_current.sql:
```diff
@@ -2,127 +2,148 @@
 -- Universe (Current Year)" open data asset.
 -- Some columns from the feeder view may not be present in this view.
 
+WITH deleted AS (
+    SELECT
+        CONCAT(parid, taxyr) AS row_id,
+        TRUE AS ":deleted" -- noqa
+    FROM {{ source('iasworld', 'pardat') }}
+    WHERE
+        deactivat IS NOT NULL
+        OR class = '999'
+),
```
We need to be able to tell socrata which rows to delete, if they're present on the open data portal. Unfortunately, these rows _don't_ exists in our athena `open_data` (or `default`) views, so we need to grab them from iasworld tables.

I'm hopeful we can turn this into a macro, but it might be rather specific for some views.

jeancochrane replied:
**[Question, non-blocking]** Is `:deleted` a Socrata-specific syntax, or is that just an arbitrary identifier we're using to describe this field?

wrridgeway commented on dbt/models/open_data/open_data.vw_parcel_universe_current.sql:
```diff
@@ -2,127 +2,150 @@
 -- Universe (Current Year)" open data asset.
 -- Some columns from the feeder view may not be present in this view.
 
+WITH deleted AS (
+    SELECT
+        parid,
+        taxyr,
+        TRUE AS ":deleted" -- noqa
+    FROM {{ source('iasworld', 'pardat') }}
+    WHERE
+        deactivat IS NOT NULL
+        OR class = '999'
+),
```
This CTE and the join at the bottom are all that's new.

jeancochrane commented on dbt/models/open_data/open_data.vw_parcel_universe_current.sql:
```diff
@@ -2,127 +2,148 @@
-    misc_subdivision_data_year
-FROM {{ ref('default.vw_pin_universe') }}
-WHERE year = (SELECT MAX(year) FROM {{ ref('default.vw_pin_universe') }})
+    COALESCE(CONCAT(feeder.pin, CAST(feeder.year AS VARCHAR)), deleted.row_id)
+        AS row_id,
+    feeder.*,
+    deleted.":deleted" -- noqa
+FROM feeder
+FULL OUTER JOIN deleted
+    ON
+    CONCAT(feeder.pin, CAST(feeder.year AS VARCHAR)) = deleted.row_id
```
**[Suggestion, optional]** I bet we could wrap a lot of this logic in a macro such that all we need to edit for each view is to define a `feeder` subquery and then pass it into the macro to get both the `deleted` CTE and the join. Happy to talk this through if more details would be helpful.

jeancochrane commented on dbt/models/open_data/open_data.vw_parcel_universe_current.sql:
```diff
@@ -2,127 +2,148 @@
+        TRUE AS ":deleted" -- noqa
+    FROM {{ source('iasworld', 'pardat') }}
+    WHERE
+        deactivat IS NOT NULL
```
**[Question, non-blocking]** One concern I have here: Due to the append-only nature of the iasWorld data model, isn't it expected that all records will have a `deactivated` representation whenever we make any edit? Do we need a more sophisticated way of querying deleted records?

wrridgeway replied:
I don't _think_ this is the case. I think that's what `cur` is for. But we should confirm.

wrridgeway replied:
```
select count(*) from iasworld.pardat
where cur = 'Y' and deactivat is not null
```

returns 0 rows.

wrridgeway replied:
and it works historically, as well:

12162130180000 | 2025 | D | ########
-- | -- | -- | --
12162130180000 | 2024 | Y |  
12162130180000 | 2023 | Y |  
12162130180000 | 2022 | Y |  
12162130180000 | 2021 | Y |  
12162130180000 | 2020 | Y |  
12162130180000 | 2019 | Y |  
12162130180000 | 2018 | Y |  
12162130180000 | 2017 | Y |  
12162130180000 | 2016 | Y |  
12162130180000 | 2015 | Y |  
12162130180000 | 2014 | Y |  
12162130180000 | 2013 | Y |  
12162130180000 | 2012 | Y |  
12162130180000 | 2011 | Y |  
12162130180000 | 2010 | Y |  
12162130180000 | 2009 | Y |  
12162130180000 | 2008 | Y |  
12162130180000 | 2007 | Y |  
12162130180000 | 2006 | Y |  
12162130180000 | 2005 | Y |  
12162130180000 | 2004 | Y |  
12162130180000 | 2003 | Y |  
12162130180000 | 2002 | Y |  
12162130180000 | 2001 | Y |  
12162130180000 | 2000 | Y |  
12162130180000 | 1999 | Y |  

jeancochrane reviewed (commented):
Nice start! This approach is less messy than I expected, especially if we can encapsulate some of the feeder/deleted logic in a reusable macro.

wrridgeway commented:
`:deleted` is what [socrata expects](https://dev.socrata.com/publishers/soda-producer/upsert).

wrridgeway commented on dbt/models/open_data/open_data.vw_parcel_status.sql:
This view _should_ include 999s in `default`, but since we don't plan on passing that indicator through to the open data portal, I didn't let them into the asset that feeds it.

wrridgeway commented:
<details>

<summary>vw_assessed_value</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_assessed_value
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_assessed_value
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| :deleted | year | new     | old     | match |
|----------|------|---------|---------|-------|
| TRUE     |      | 20203   |         |       |
|          | 2023 | 1864161 | 1864161 | TRUE  |
|          | 2009 | 1849650 | 1849650 | TRUE  |
|          | 2005 | 1740963 | 1740963 | TRUE  |
|          | 2016 | 1863996 | 1863996 | TRUE  |
|          | 1999 | 1620798 | 1620798 | TRUE  |
|          | 2019 | 1864102 | 1864102 | TRUE  |
|          | 2024 | 1864107 | 1864107 | TRUE  |
|          | 2003 | 1693139 | 1693139 | TRUE  |
|          | 2007 | 1804217 | 1804217 | TRUE  |
|          | 2018 | 1864588 | 1864588 | TRUE  |
|          | 2021 | 1864589 | 1864589 | TRUE  |
|          | 2011 | 1861008 | 1861008 | TRUE  |
|          | 2015 | 1862756 | 1862756 | TRUE  |
|          | 2008 | 1832617 | 1832617 | TRUE  |
|          | 2020 | 1864035 | 1864035 | TRUE  |
|          | 2012 | 1862181 | 1862181 | TRUE  |
|          | 2025 | 1863560 | 1863560 | TRUE  |
|          | 2014 | 1862170 | 1862170 | TRUE  |
|          | 2013 | 1861929 | 1861929 | TRUE  |
|          | 2010 | 1859244 | 1859244 | TRUE  |
|          | 2004 | 1716214 | 1716214 | TRUE  |
|          | 2006 | 1771313 | 1771313 | TRUE  |
|          | 2000 | 1637079 | 1637079 | TRUE  |
|          | 2001 | 1655168 | 1655168 | TRUE  |
|          | 2002 | 1673530 | 1673530 | TRUE  |
|          | 2017 | 1864621 | 1864621 | TRUE  |
|          | 2022 | 1865013 | 1865013 | TRUE  |

</details>

<details>

<summary>vw_parcel_address</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_address
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_parcel_address
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| :deleted | year | new     | old     | match |
|----------|------|---------|---------|-------|
| TRUE     |      | 20203   |         |       |
|          | 2021 | 1864589 | 1864589 | TRUE  |
|          | 2019 | 1864102 | 1864102 | TRUE  |
|          | 2006 | 1771314 | 1771314 | TRUE  |
|          | 2011 | 1861008 | 1861008 | TRUE  |
|          | 2001 | 1655398 | 1655398 | TRUE  |
|          | 2009 | 1849650 | 1849650 | TRUE  |
|          | 1999 | 1620990 | 1620990 | TRUE  |
|          | 2007 | 1804219 | 1804219 | TRUE  |
|          | 2018 | 1864590 | 1864590 | TRUE  |
|          | 2012 | 1862181 | 1862181 | TRUE  |
|          | 2022 | 1865016 | 1865016 | TRUE  |
|          | 2015 | 1862756 | 1862756 | TRUE  |
|          | 2013 | 1861935 | 1861935 | TRUE  |
|          | 2016 | 1863996 | 1863996 | TRUE  |
|          | 2003 | 1693153 | 1693153 | TRUE  |
|          | 2002 | 1673735 | 1673735 | TRUE  |
|          | 2008 | 1832618 | 1832618 | TRUE  |
|          | 2025 | 1863560 | 1863560 | TRUE  |
|          | 2014 | 1862170 | 1862170 | TRUE  |
|          | 2017 | 1864621 | 1864621 | TRUE  |
|          | 2020 | 1864035 | 1864035 | TRUE  |
|          | 2000 | 1637269 | 1637269 | TRUE  |
|          | 2024 | 1864107 | 1864107 | TRUE  |
|          | 2010 | 1859244 | 1859244 | TRUE  |
|          | 2004 | 1716218 | 1716218 | TRUE  |
|          | 2023 | 1864161 | 1864161 | TRUE  |
|          | 2005 | 1740965 | 1740965 | TRUE  |

</details>

<details>

<summary>vw_parcel_status</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_status
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_parcel_status
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 1  | true     | 2025 | 609     | 1864169 | false |
| 2  |          | 2025 | 1863560 | 1864169 | false |
| 3  |          | 2024 | 1864107 | 1864107 | true  |
| 4  | true     | 2023 | 1       | 1864162 | false |
| 5  |          | 2023 | 1864161 | 1864162 | false |
| 6  |          | 2022 | 1865016 | 1865018 | false |
| 7  | true     | 2022 | 2       | 1865018 | false |
| 8  |          | 2021 | 1864589 | 1864589 | true  |
| 9  |          | 2020 | 1864035 | 1864035 | true  |
| 10 |          | 2019 | 1864102 | 1864102 | true  |
| 11 |          | 2018 | 1864590 | 1864590 | true  |
| 12 |          | 2017 | 1864621 | 1864621 | true  |
| 13 |          | 2016 | 1863996 | 1863996 | true  |
| 14 |          | 2015 | 1862756 | 1862756 | true  |
| 15 |          | 2014 | 1862170 | 1862170 | true  |
| 16 |          | 2013 | 1861935 | 1861935 | true  |
| 17 |          | 2012 | 1862181 | 1862181 | true  |
| 18 |          | 2011 | 1861008 | 1861008 | true  |
| 19 |          | 2010 | 1859244 | 1859244 | true  |
| 20 |          | 2009 | 1849650 | 1849650 | true  |
| 21 |          | 2008 | 1832618 | 1832618 | true  |
| 22 |          | 2007 | 1804219 | 1804219 | true  |
| 23 |          | 2006 | 1771314 | 1771314 | true  |
| 24 |          | 2005 | 1740965 | 1740965 | true  |
| 25 |          | 2004 | 1716218 | 1716218 | true  |
| 26 |          | 2003 | 1693153 | 1693153 | true  |
| 27 |          | 2002 | 1673735 | 1673735 | true  |
| 28 |          | 2001 | 1655398 | 1655398 | true  |
| 29 |          | 2000 | 1637269 | 1637269 | true  |
| 30 |          | 1999 | 1620990 | 1620990 | true  |
| 31 | true     |      | 19591   |

</details>

<details>

<summary>vw_parcel_universe_current</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_current
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_parcel_universe_current
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| # | :deleted | year | new     | old     | match |
|---|----------|------|---------|---------|-------|
| 1 |          | 2025 | 1863560 | 1863560 | true  |
| 2 | true     |      | 20203   |

</details>

<details>

<summary>vw_parcel_universe_historical</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_parcel_universe_historical
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_parcel_universe_historical
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 10 | true     |      | 20203   |         |       |
| 1  |          | 2011 | 1861008 | 1861008 | true  |
| 2  |          | 2012 | 1862181 | 1862181 | true  |
| 3  |          | 2007 | 1804219 | 1804219 | true  |
| 4  |          | 2016 | 1863996 | 1863996 | true  |
| 5  |          | 2019 | 1864102 | 1864102 | true  |
| 6  |          | 2003 | 1693153 | 1693153 | true  |
| 7  |          | 2015 | 1862756 | 1862756 | true  |
| 8  |          | 2010 | 1859244 | 1859244 | true  |
| 9  |          | 2008 | 1832618 | 1832618 | true  |
| 11 |          | 2002 | 1673735 | 1673735 | true  |
| 12 |          | 2020 | 1864035 | 1864035 | true  |
| 13 |          | 2014 | 1862170 | 1862170 | true  |
| 14 |          | 2006 | 1771314 | 1771314 | true  |
| 15 |          | 2005 | 1740965 | 1740965 | true  |
| 16 |          | 2022 | 1865016 | 1865016 | true  |
| 17 |          | 2018 | 1864590 | 1864590 | true  |
| 18 |          | 2017 | 1864621 | 1864621 | true  |
| 19 |          | 2004 | 1716218 | 1716218 | true  |
| 20 |          | 1999 | 1620990 | 1620990 | true  |
| 21 |          | 2025 | 1863560 | 1863560 | true  |
| 22 |          | 2001 | 1655398 | 1655398 | true  |
| 23 |          | 2013 | 1861935 | 1861935 | true  |
| 24 |          | 2023 | 1864161 | 1864161 | true  |
| 25 |          | 2009 | 1849650 | 1849650 | true  |
| 26 |          | 2000 | 1637269 | 1637269 | true  |
| 27 |          | 2024 | 1864107 | 1864107 | true  |
| 28 |          | 2021 | 1864589 | 1864589 | true  |

</details>

<details>

<summary>vw_property_tax_exempt_parcel</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_property_tax_exempt_parcel
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_property_tax_exempt_parcel
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| # | :deleted | year | new   | old   | match |
|---|----------|------|-------|-------|-------|
| 1 | true     |      | 20203 |       |       |
| 2 |          | 2024 | 93509 | 93509 | true  |
| 3 |          | 2025 | 93409 | 93409 | true  |
| 4 |          | 2022 | 92416 | 92416 | true  |
| 5 |          | 2023 | 92481 | 92481 | true  |

</details>

<details>

<summary>vw_res_condo_unit_char</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_res_condo_unit_char
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_res_condo_unit_char
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| :deleted | year | new    | old    | match |
|----------|------|--------|--------|-------|
| TRUE     |      | 20203  |        |       |
|          | 2003 | 300884 | 300884 | TRUE  |
|          | 2001 | 266830 | 266830 | TRUE  |
|          | 2021 | 458980 | 458980 | TRUE  |
|          | 2011 | 457166 | 457166 | TRUE  |
|          | 2010 | 454939 | 454939 | TRUE  |
|          | 2012 | 458194 | 458194 | TRUE  |
|          | 2007 | 402508 | 402508 | TRUE  |
|          | 2015 | 458355 | 458355 | TRUE  |
|          | 2022 | 459126 | 459126 | TRUE  |
|          | 2009 | 445790 | 445790 | TRUE  |
|          | 2017 | 459275 | 459275 | TRUE  |
|          | 2006 | 370734 | 370734 | TRUE  |
|          | 2023 | 457640 | 457640 | TRUE  |
|          | 2014 | 458049 | 458049 | TRUE  |
|          | 2016 | 458847 | 458847 | TRUE  |
|          | 2018 | 459473 | 459473 | TRUE  |
|          | 2005 | 342996 | 342996 | TRUE  |
|          | 2013 | 457812 | 457812 | TRUE  |
|          | 2002 | 283450 | 283450 | TRUE  |
|          | 2019 | 458673 | 458673 | TRUE  |
|          | 2024 | 457315 | 457315 | TRUE  |
|          | 2000 | 250302 | 250302 | TRUE  |
|          | 2004 | 321082 | 321082 | TRUE  |
|          | 2008 | 429185 | 429185 | TRUE  |
|          | 1999 | 235795 | 235795 | TRUE  |
|          | 2025 | 457216 | 457216 | TRUE  |
|          | 2020 | 458449 | 458449 | TRUE  |

</details>

<details>

<summary>vw_sf_mf_improvement_char</summary>

```
with new as (
select ":deleted", year, count(*) as new
from z_ci_fix_open_data_row_deletion_open_data.vw_sf_mf_improvement_char
group by year, ":deleted"
),

old AS (
select year, count(*) as old
from open_data.vw_sf_mf_improvement_char
group by year
)

select
    new.":deleted",
    new.year,
    new.new,
    old.old,
    new.new = old.old as match
from new
full outer join old on new.year = old.year
```

| #  | :deleted | year | new     | old     | match |
|----|----------|------|---------|---------|-------|
| 12 | true     |      | 20203   |         |       |
| 1  |          | 2007 | 1098544 | 1098544 | true  |
| 2  |          | 2008 | 1099960 | 1099960 | true  |
| 3  |          | 2001 | 1078951 | 1078951 | true  |
| 4  |          | 2003 | 1085834 | 1085834 | true  |
| 5  |          | 2017 | 1098637 | 1098637 | true  |
| 6  |          | 2016 | 1098108 | 1098108 | true  |
| 7  |          | 2004 | 1089762 | 1089762 | true  |
| 8  |          | 2020 | 1099085 | 1099085 | true  |
| 9  |          | 2010 | 1099597 | 1099597 | true  |
| 10 |          | 2011 | 1099010 | 1099010 | true  |
| 11 |          | 2002 | 1081904 | 1081904 | true  |
| 13 |          | 2000 | 1077974 | 1077974 | true  |
| 14 |          | 2024 | 1102434 | 1102438 | false |
| 15 |          | 2018 | 1098786 | 1098786 | true  |
| 16 |          | 2023 | 1101126 | 1101126 | true  |
| 17 |          | 2021 | 1099631 | 1099631 | true  |
| 18 |          | 2015 | 1098115 | 1098115 | true  |
| 19 |          | 2019 | 1098966 | 1098966 | true  |
| 20 |          | 2022 | 1100239 | 1100239 | true  |
| 21 |          | 2009 | 1099466 | 1099466 | true  |
| 22 |          | 2005 | 1093404 | 1093404 | true  |
| 23 |          | 2014 | 1097705 | 1097705 | true  |
| 24 |          | 1999 | 1059702 | 1059702 | true  |
| 25 |          | 2012 | 1098752 | 1098752 | true  |
| 26 |          | 2025 | 1102309 | 1102315 | false |
| 27 |          | 2013 | 1098025 | 1098025 | true  |
| 28 |          | 2006 | 1096341 | 1096341 | true  |

</details>

wrridgeway commented on dbt/macros/open_data_rows_to_delete.sql:
```diff
@@ -0,0 +1,43 @@
+            select
+                pdat.parid || cast(ddat.card as varchar) || pdat.taxyr as row_id,
+                true as ":deleted"  -- noqa
+            from {{ source("iasworld", "pardat") }} as pdat
+            inner join
+                {{ source("iasworld", "dweldat") }} as ddat
```
our res chars view is actually based on `dwelda`t, but I do not trust that table as a source of which pins have been deactivated or switched to class 999. We have found class mismatches between `pardat` and the other tables in the past and `pardat` is usually the table with the correct class.

wrridgeway commented:
This is currently [working](https://github.com/ccao-data/data-architecture/actions/runs/14890918744/job/41822500607) as expected:

```
select count(*)
from z_ci_fix_open_data_row_deletion_open_data.vw_sf_mf_improvement_char
where year = 2025
and ":deleted" is null
```

`1102311`

![image](https://github.com/user-attachments/assets/31b20ab2-cdbf-406b-a2cc-544b0063821e)

wrridgeway commented:
I am having a heck of a time getting this to work properly for the single family asset. All the others seem to be working just fine.  [Here's](https://datacatalog.cookcountyil.gov/resource/x54s-btds.json?$query=SELECT%0A%20%20%60pin%60%2C%0A%20%20%60year%60%2C%0A%20%20%60card%60%2C%0A%20%20%60class%60%2C%0A%20%20%60township_code%60%2C%0A%20%20%60tieback_key_pin%60%2C%0A%20%20%60tieback_proration_rate%60%2C%0A%20%20%60card_proration_rate%60%2C%0A%20%20%60cdu%60%2C%0A%20%20%60pin_is_multicard%60%2C%0A%20%20%60pin_num_cards%60%2C%0A%20%20%60pin_is_multiland%60%2C%0A%20%20%60pin_num_landlines%60%2C%0A%20%20%60char_yrblt%60%2C%0A%20%20%60char_bldg_sf%60%2C%0A%20%20%60char_land_sf%60%2C%0A%20%20%60char_beds%60%2C%0A%20%20%60char_rooms%60%2C%0A%20%20%60char_fbath%60%2C%0A%20%20%60char_hbath%60%2C%0A%20%20%60char_frpl%60%2C%0A%20%20%60char_type_resd%60%2C%0A%20%20%60char_cnst_qlty%60%2C%0A%20%20%60char_apts%60%2C%0A%20%20%60char_attic_fnsh%60%2C%0A%20%20%60char_gar1_att%60%2C%0A%20%20%60char_gar1_area%60%2C%0A%20%20%60char_gar1_size%60%2C%0A%20%20%60char_gar1_cnst%60%2C%0A%20%20%60char_attic_type%60%2C%0A%20%20%60char_bsmt%60%2C%0A%20%20%60char_ext_wall%60%2C%0A%20%20%60char_heat%60%2C%0A%20%20%60char_repair_cnd%60%2C%0A%20%20%60char_bsmt_fin%60%2C%0A%20%20%60char_roof_cnst%60%2C%0A%20%20%60char_use%60%2C%0A%20%20%60char_site%60%2C%0A%20%20%60char_ncu%60%2C%0A%20%20%60char_renovation%60%2C%0A%20%20%60char_porch%60%2C%0A%20%20%60char_air%60%2C%0A%20%20%60char_tp_plan%60%0AWHERE%20caseless_one_of(%60pin%60%2C%20%2205063030010000%22)%20AND%20%60year%60%20IN%20(%222025%22)%0AORDER%20BY%20%60year%60%20DESC%20NULL%20LAST) what's currently on the open data portal for an example PIN.

If you run this query in athena you can see that cards 1 and 3 are marked for deletion:
```
select *
from z_ci_fix_open_data_row_deletion_open_data.vw_sf_mf_improvement_char
where substr(row_id, 1, 14) = '05063030010000' and year = 2025
```
But [running the workflow](https://github.com/ccao-data/data-architecture/actions/runs/14930842629/job/41946442645) doesn't delete anything. I'm kind of stumped on this one. I _think_ for some reason the deleted rows aren't actually being sent?

wrridgeway commented on dbt/macros/open_data_rows_to_delete.sql:
```diff
@@ -0,0 +1,65 @@
+                    pdat.taxyr as year,
+                    true as ":deleted"
+                from {{ source("iasworld", "pardat") }} as pdat
+                {%- if own == true %}
```
I don't think we need to worry about making the `own` or `condo` parameters work with the `card` parameter. There should never be a case where they are all used together.

jeancochrane reviewed (approved):
Very nice! This isn't even as bad as I thought it would be.

I do think we should update the open data row count tests sooner rather than later. Do you want to do that as part of this PR, or as a fast follow-up?

wrridgeway commented on dbt/macros/open_data_rows_to_delete.sql:
```diff
@@ -0,0 +1,65 @@
+is different for different feeder views. The macro takes arguments to specify
+how to construct the approriate universe of rows to purge.
+*/
+{%- macro open_data_rows_to_delete(
```
Agreed.

wrridgeway commented on dbt/models/open_data/open_data.vw_parcel_universe_current.sql:
```diff
@@ -3,98 +3,96 @@
 -- Some columns from the feeder view may not be present in this view.
 
 SELECT
```
I think this is a good suggestion, and I'll open an issue to address it.

wrridgeway commented:
I'll update the tests as a fast follow-up.
